csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3620 @@
+      iel_promotion_map[out_groups.vector()[i]] = replay_out_ids[i];
+    }
+  }
+
+  // Map from an exact iter domain group, to all the exact iter domain groups it
+  // covers
+  std::unordered_map<IdGroup, IdGroups> exact_covered_ids;
+
+  for (auto id_group :
+       idGraph(IdMappingMode::EXACT).disjointIdSets().disjointSets()) {
+    // Initialize inputs
+    if (idGraph(IdMappingMode::EXACT).uniqueDefinitions(id_group).empty()) {
+      exact_covered_ids[id_group] = {id_group};
+    }
+
+    // Initialize rfactor groups
+    if (std::any_of(id_group->begin(), id_group->end(), [&](IterDomain* id) {
+          return view_rfactor_ids_.find(id) != view_rfactor_ids_.end();
+        })) {
+      exact_covered_ids[id_group] = {id_group};
+    }
+
+    // Initialize broadcast groups to empty
+    if (std::any_of(id_group->begin(), id_group->end(), [&](IterDomain* id) {
+          return id->isBroadcast();
+        })) {
+      exact_covered_ids[id_group] = {};
+    }
+  }
+
+  IdGraphStmtSort exact_stmt_sort(idGraph(IdMappingMode::EXACT));
+
+  for (auto exact_expr : exact_stmt_sort.exprs()) {
+    auto input_groups = idGraph(IdMappingMode::EXACT).inputGroups(exact_expr);
+
+    IdGroups covered;
+    for (auto inp_group : input_groups) {
+      covered.pushBack(exact_covered_ids.at(inp_group));
+    }
+
+    for (auto output_group :
+         idGraph(IdMappingMode::EXACT).outputGroups(exact_expr)) {
+      exact_covered_ids[output_group] = covered;
+    }
+  }
+
+  // Loop promotion map is to prepare for IterDomain replays. Since these
+  // replays will modify the loop map, we operate on a copy of the loop map,
+  // not the original one.
+
+  auto loop_graph_copy = idGraph(IdMappingMode::LOOP);
+  std::unordered_map<IdGroup, IterDomain*> loop_graph_copy_promotion_map;
+
+  for (auto loop_group : loop_graph_copy.disjointIdSets().disjointSets()) {
+    if (loop_group->size() == 1) {
+      loop_graph_copy_promotion_map[loop_group] = loop_group->front();
+      continue;
+    }
+
+    // We need to check the exact groups the terminal id's are in, but for
+    // promotion we want an iter domain within the loop group. Since exact
+    // group can traverse loop group boundaires, save a vector of the group
+    // and the iter domain.
+    std::vector<std::pair<IdGroup, IterDomain*>> exact_promoted_terminal_ids;
+    for (auto loop_id : *loop_group) {
+      if (terminal_loop_ids.has(loop_id)) {
+        auto iel_set_pair =
+            intersection_exact_loop_graph.disjointIdSet(loop_id);
+        TORCH_INTERNAL_ASSERT(iel_set_pair.second);
+        auto iel_group = iel_set_pair.first;
+        auto iel_promo_it = iel_promotion_map.find(iel_group);
+        if (iel_promo_it == iel_promotion_map.end()) {
+          auto promo_id_exact_it =
+              idGraph(IdMappingMode::EXACT).disjointIdSet(loop_id);
+          TORCH_INTERNAL_ASSERT(promo_id_exact_it.second);
+          exact_promoted_terminal_ids.push_back(
```

Comment:
In this case, `loop_id` is not promoted. Why does this need to be tracked?

Reply:
`exact_promoted_terminal_ids` is a bit misleading of a name. We're just looking for the candidates for concrete ID's.

The concrete ID could be:
(1) The only iter domain in the group if the group only has one iter domain
(2) An iter domain that's simply within the loop group if no promotion happened
(3) One of the terminal promoted id's that had to be generated

Reply:
Oh, I see. I didn't realize this part is for finding concrete IDs.